mislabeled/probe/_linear.py

```diff
@@ -46,6 +47,99 @@
 class LinearModel(NamedTuple):
     coef: np.ndarray
     intercept: np.ndarray
+    loss: str
+    regul: float
+
+    def decision_function(self, X):
+        y_linear = X @ self.coef
+        if self.intercept is not None:
+            y_linear += self.intercept
+        if y_linear.ndim == 1:
+            y_linear = y_linear[:, None]
+        return y_linear
+
+    def predict_proba(self, X):
+        y_linear = self.decision_function(X)
+        if self.loss == "l2":
+            return y_linear
+        elif self.loss == "log_loss":
+            if self._is_binary():
+                return expit(y_linear)
+            else:
+                return softmax(y_linear, axis=1)
+        else:
+            raise NotImplementedError()
+
+    def grad_y(self, X, y):
+        # gradients w.r.t. the output of the linear op, i.e. the logit
+        # in the logistic model
+        y_linear = self.decision_function(X)
+        if self.loss == "l2":
+            if y.ndim == 1:
+                y = y[:, None]
+            dl_dy = 2 * (y - y_linear)
+
+        elif self.loss == "log_loss":
+            y_linear = self.decision_function(X)
+            if self._is_binary():
+                dl_dy = y[:, None] - expit(y_linear)
+            else:
+                dl_dy = -softmax(y_linear, axis=1)
+                dl_dy[np.arange(y.shape[0]), y] += 1
+
+        else:
+            raise NotImplementedError()
+        return dl_dy
+
+    def grad_p(self, X, y):
+        # gradients w.r.t. the parameters (weight, intercept)
+        dl_dy = self.grad_y(X, y)
+
+        X_p = X
+        if self.intercept is not None:
+            X_p = np.hstack((X, np.ones((X.shape[0], 1))))
+
+        return dl_dy[:, :, None] * X_p[:, None, :]
+
+    def hessian(self, X, y):
```

Comment:
does it scale to models used in the benchmark? the previous sparse implementation was meant to do that.

Reply:
Maybe create a new issue so that we can close this pull request.

Reply:
#11